src/beast/beast/nudb/store.h

```diff
@@ -414,229 +413,208 @@ basic_store<Hasher, File>::close()
     }
 }
 
-template <class Hasher, class File>
-template <class BufferFactory>
+template <class Hasher, class Codec, class File>
+template <class Handler>
 bool
-basic_store<Hasher, File>::fetch (
-    void const* key, BufferFactory&& bf)
+store<Hasher, Codec, File>::fetch (
+    void const* key, Handler&& handler)
 {
     using namespace detail;
     rethrow();
-    std::size_t offset;
-    std::size_t size;
-    blockbuf buf(s_->b);
-    bucket tmp (s_->kh.key_size,
-        s_->kh.block_size, buf.get());
+    auto const h = hash<Hasher>(
+        key, s_->kh.key_size, s_->kh.salt);
+    shared_lock_type m (m_);
     {
-        auto const h = hash<Hasher>(
-           key, s_->kh.key_size, s_->kh.salt);
-        shared_lock_type m (m_,
-            boost::defer_lock);
-        m.lock();
+        auto iter = s_->p1.find(key);
+        if (iter == s_->p1.end())
         {
-            typename pool::iterator iter;
-            iter = s_->p1.find(key);
-            if (iter != s_->p1.end())
-            {
-                void* const b = bf(
-                    iter->first.size);
-                if (b == nullptr)
-                    return false;
-                std::memcpy (b,
-                    iter->first.data,
-                        iter->first.size);
-                return true;
-            }
             iter = s_->p0.find(key);
-            if (iter != s_->p0.end())
-            {
-                void* const b = bf(
-                    iter->first.size);
-                if (b == nullptr)
-                    return false;
-                std::memcpy (b,
-                    iter->first.data,
-                        iter->first.size);
-                return true;
-            }
-        }
-        auto const n = bucket_index(
-            h, buckets_, modulus_);
-        auto const iter = s_->c1.find(n);
-        if (iter != s_->c1.end())
-        {
-            auto const result =
-                iter->second.find(key);
-            if (result.second)
-            {
-                offset = result.first.offset;
-                size = result.first.size;
-                goto found;
-            }
-            // VFALCO Audit for concurrency
-            auto spill = iter->second.spill();
-            m.unlock();
-            while (spill)
-            {
-                tmp.read(s_->df, spill);
-                auto const result = tmp.find(key);
-                if (result.second)
-                {
-                    offset = result.first.offset;
-                    size = result.first.size;
-                    goto found;
-                }
-                spill = tmp.spill();
-            }
-            return false;
+            if (iter == s_->p0.end())
+                goto next;
```

Comment:
One way to address this is to create a private helper function:
````
auto
find_in_pool(void const* key)
{
    auto iter = s_->p1.find(key);
    if (iter == s_->p1..end())
        iter = s_->p0.find(key);
    return iter;
}
````
This can then be used here like this:
````
    shared_lock_type m (m_);
    {
        auto iter = find_in_pool(key);
        if (iter != pool_end())  // pool_end() == s_->p0.end()
        {
            buffer buf;
            auto const result =
                s_->codec.decompress(
                iter->first.data,
                    iter->first.size, buf);
            handler(result.first, result.second);
            return true;
        }
    }
````
This same helper can also be used on line 479 of this same file:
````
        shared_lock_type m (m_);
        if (find_in_pool(key) != pool_end())
            return false;
````
Disclaimer:  untested code.